acceptance: add password authentication tests

asubiotto opened the pull request:
Necessary to test cli integration with server password authentication.
Also refactored some tests to be able to use an empty environment. The
default environment would result in the client providing the PGSSLCERT
to the server.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10449)
<!-- Reviewable:end -->

Commit:
acceptance: add password authentication tests

Necessary to test cli integration with server password authentication.
Also refactored some tests to be able to use an empty environment. The
default environment would result in the client providing the PGSSLCERT
to the server.

tamird commented:
Reviewed 3 of 4 files at r1.
Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KVhHG5EmtNiZf-lEEQZ:b-k79mcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
> ```Go
> )
> 
> func testDocker(t *testing.T, num int32, name string, cmd, env []string) error {
> ```

how about making this method take a `container.Config` as an argument instead of `cmd` and `env`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 3 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-:-KVhNbJ1fjviJX1ushlG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KVjBLCuR2m8Q6qPZ2eK:b-6dkx4c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how about making this method take a `container.Config` as an argument instead of `cmd` and `env`?
</details>
I'm in favor of introducing a more general interface, but do we have tests that would use that already?
Also I'm not fond of having the callers of this function have to specify the Docker image explicitly.

---

*[pkg/cli/interactive_tests/test_secure.tcl, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVjCC8nNCZ1h5PaXuJX:-KVjCC8nNCZ1h5PaXuJY:b-eck0h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf89380421c4c5a2b9b8cdf67d5aa77f3625c1b6/pkg/cli/interactive_tests/test_secure.tcl#L12)):*
> ```Tcl
> start_secure_server $argv
> 
> spawn $argv node ls --ca-cert=/certs/ca.crt --cert=/certs/node.crt --key=/certs/node.key
> ```

You're making this test run more heavyweight than it needs by using only "spawn". If I were you I would spawn a shell just once then use "send" to launch the cockroach commands, and expect the shell's prompt between commands.

---

*[pkg/cli/interactive_tests/test_secure.tcl, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVjBjo-d3BUs62LkmzZ:-KVjBjo0QyVbYTtJ9khW:bb8cznu) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/cli/interactive_tests/test_secure.tcl#L36)):*
> ```Tcl
> eexpect "user nonexistent does not exist"
> 
> # Root can only authenticate using certificate authentication.
> ```

Wait what is happening here. The error message is "user root cannot authenticate using a password" but the command does not specify a password. 
I think the error message is not properly informative, it should probably be "user root must authenticate using a client certificate" instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KVjVPXw07NXvmKb5Uby:bgofnh5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> I'm in favor of introducing a more general interface, but do we have tests that would use that already?
> Also I'm not fond of having the callers of this function have to specify the Docker image explicitly.
</details>
We could have our own acceptance Config in which ```name```, ```cmd```, and ```env``` are specified and keep the specification of the Docker image in here. Does that sound good?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->

knz commented:
Yep, sounds good!

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KVzPGNnuUNp7Q7yBoD8:bi98gse) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary>
> We could have our own acceptance Config in which ```name```, ```cmd```, and ```env``` are specified and keep the specification of the Docker image in here. Does that sound good?
</details>
I don't think that buys us anything, tbh. If anything, I would rather we separate the images used for these tests. The CLI tests only want `tcl`, while the other tests want a bunch of language-specific postgres drives.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->

asubiotto commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KW3P7LQnuvuETdYCqOC:broppoz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I don't think that buys us anything, tbh. If anything, I would rather we separate the images used for these tests. The CLI tests only want `tcl`, while the other tests want a bunch of language-specific postgres drives.
</details>
Changed to container.Config and have the image specified in both ```cli_test``` and ```reference_test```. Other tests use the default container through ```testDocker{Success, Fail}```

---

*[pkg/cli/interactive_tests/test_secure.tcl, line 12 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVjCC8nNCZ1h5PaXuJX:-KW3PWFQzN5yU7CE3IEK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf89380421c4c5a2b9b8cdf67d5aa77f3625c1b6/pkg/cli/interactive_tests/test_secure.tcl#L12)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> You're making this test run more heavyweight than it needs by using only "spawn". If I were you I would spawn a shell just once then use "send" to launch the cockroach commands, and expect the shell's prompt between commands.
</details>
Done.

---

*[pkg/cli/interactive_tests/test_secure.tcl, line 36 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVjBjo-d3BUs62LkmzZ:-KW3PbaBo4dyeJQRFysg:b-y2c5tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/cli/interactive_tests/test_secure.tcl#L36)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Wait what is happening here. The error message is "user root cannot authenticate using a password" but the command does not specify a password. 
> I think the error message is not properly informative, it should probably be "user root must authenticate using a client certificate" instead.
</details>
I agree with you. Changed it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KW3RtA05FQ2_cqhXaJs:b-l5eosd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary>
> Changed to container.Config and have the image specified in both ```cli_test``` and ```reference_test```. Other tests use the default container through ```testDocker{Success, Fail}```
</details>
@tamird I agree the images could be separated, but perhaps that deserves another PR? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-:-KW3SBQLT3YhMPjSzyPV:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 6 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/util_test.go, line 468 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KVhHG5EmtNiZf-lEEQY:-KW3a6U0ZK0rMBhq3ECI:b-6scift) ([raw file](https://github.com/cockroachdb/cockroach/blob/6610f09676a043836522e348931a9af880b92714/pkg/acceptance/util_test.go#L468)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> @tamird I agree the images could be separated, but perhaps that deserves another PR? 
</details>
Yep, it certainly deserves another PR.

---

*[pkg/acceptance/util_test.go, line 466 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449#-KW3amLoiH8AYEFVauD0:-KW3amLoiH8AYEFVauD1:bhzgbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eaf2cef8cd3ffce3b785b046c677093562d6416/pkg/acceptance/util_test.go#L466)):*
> ```Go
> )
> 
> var defaultContainerConfig = container.Config{
> ```

this should be a function, not a global variable. we've had issues with accidental modification of these global in the past.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10449)*
<!-- Sent from Reviewable.io -->
